File: javascript/grade-school/grade-school.js
export class GradeSchool {
  constructor() {
    this.grades = {}
  }

  add(student, grade) {
    this.grades[grade] = [...this.students(grade), student].sort()
  }

  roster() {
    return Object.keys(this.grades)
      .reduce((object, key) => ({
        ...object,
        [key]: [...this.grades[key]]
      }), {})
  }

  grade(g) {
    return [...this.students(g)]
  }

  students(grade) {
    return grade in this.grades ?
      this.grades[grade] :
      [];
  }
}

File: ruby/grade-school/grade_school.rb
class School
  def initialize
    @grades = Hash.new { |hash, key| hash[key] = [] }
  end

  def students(grade)
    grades[grade]
  end

  def add(name, grade)
    grades[grade] << name
    grades[grade].sort!
  end

  def students_by_grade
    grades
      .keys
      .sort
      .each_with_object([])
      .map { |grade| { grade: grade, students: students(grade)} }
  end

  private

  attr_reader :grades
end
